tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/MeanTensor.java

```diff
@@ -0,0 +1,186 @@
+  }
+  /**
+   * Creates a MeanTensor metric
+   *
+   * @param tf the TensorFlow ops
+   * @param name the name of this metric, if null then {@link Class#getSimpleName()} is used
+   * @param seed the seed for random number generation. An initializer created with a given seed
+   *     will always produce the same random tensor for a given shape and data type.
+   * @param type the data type for the variables
+   */
+  public MeanTensor(Ops tf, String name, long seed, Class<T> type) {
+    super(tf, name, seed);
+    this.type = type;
+    this.totalName = this.getVariableName(TOTAL);
+    this.countName = this.getVariableName(COUNT);
+  }
+
+  /**
+   * Creates the Operations that initialize the total and count variables.
+   *
+   * @param shape the shape of the variables
+   * @return true if the variables need initialization, otherwise false;
+   */
+  private boolean init(Shape shape) {
+    if (!initialized) {
+      this.shape = shape;
+      Zeros<T> zeros = new Zeros<>(getTF());
+      Operand<T> zero = zeros.call(getTF().constant(shape), type);
+
+      if (total == null) {
+        total = getTF().withName(totalName).variable(zero);
+        totalInitializer = getTF().assign(total, zero);
+      }
+      if (count == null) {
+        count = getTF().withName(countName).variable(zero);
+        countInitializer = getTF().assign(count, zero);
+      }
+      this.initialized = true;
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public List<Op> updateStateList(
+      Operand<? extends TNumber> values, Operand<? extends TNumber> sampleWeights) {
+    Ops tf = getTF();
+    Operand<T> tValues = cast(tf, values, type);
+    Operand<T> tSampleWeights = null;
+    if (sampleWeights != null) tSampleWeights = cast(tf, sampleWeights, type);
+
+    boolean needsInitialization = init(values.shape());
+
+    if (!this.shape.equals(values.shape())) {
+      throw new IllegalArgumentException(
+          String.format(
+              "MeanTensor input values must always have the same shape. Expected shape (set during the first call): %s. Got %s",
+              this.shape.toString(), values.shape().toString()));
+    }
+
+    Operand<T> numValues = tf.onesLike(tValues);
+    if (tSampleWeights != null) {
+      LossTuple<T> tuple =
+          LossesHelper.squeezeOrExpandDimensions(tf, null, tValues, tSampleWeights);
+      tValues = tuple.getTarget();
+      tSampleWeights = tuple.getSampleWeights();
+      try {
+        tSampleWeights = WeightsBroadcastOps.broadcastWeights(tf, tSampleWeights, tValues);
+      } catch (IllegalArgumentException ex) {
+        int ndim = values.shape().numDimensions();
+        int weightNdim = tSampleWeights.asOutput().shape().numDimensions();
+        int[] range = new int[ndim - weightNdim];
+        for (int i = weightNdim; i < ndim; i++) {
+          range[i] = i;
+        }
+        tValues = tf.math.mean(tValues, tf.constant(range));
```

Comment:
I see that the Python code uses this same logic, but doesn't it leave `tValues` with a different shape than `numValues` and `this.total`? I don't see a test that covers this case.

Reply:
Your analysis seems right, let me test it in Python.

Reply:
There is no unit test in Python for this condition. When I run it, using Python, with the following 2 tests cases, it does not work for either.  We could either throw an error or perhaps add in some reshapes and squeezes to handle these conditions.

Test 1: values shape higher rank than weights.
```
values shape = [2,2,2]
sample_weight shape = [2]
--
num_values shape = [2]
values shape = [2]
variable shape = [2,2,2]
tensorflow.python.framework.errors_impl.InvalidArgumentError: Cannot update variable with shape [2,2,2] using a Tensor with shape [2], shapes must be equal. [Op:AssignAddVariableOp]
```

Test 2: weights shape higher rank than values
```
values shape = [2]
sample_weight shape = [2,1,1]
--
num_values shape = [2]
values shape = [2,1,2]
variable shape = [2]
tensorflow.python.framework.errors_impl.InvalidArgumentError: Cannot update variable with shape [2] using a Tensor with shape [2,1,2], shapes must be equal. [Op:AssignAddVariableOp]
```

Reply:
I'd propose we just not catch the `IllegalArgumentException` from `WeightsBroadcastOps.broadcastWeights`. It has a fine message. I'm taking this as an idea in the Python code that was never tested and never worked. We have so much work to do that I'd advocate not finishing this Python idea right now.